sdk/trace/span.go

```diff
@@ -225,9 +224,17 @@ func (s *span) End(options ...trace.SpanEndOption) {
 		return
 	}
 
+	// If the span was not properly started, stopwatch will be nil.
+	// The normal check (IsRecording) happens after end time generation,
+	// so we add a simple check here to make sure end time is generated
+	// as early as possible.
+	if s.stopwatch == nil {
+		return
+	}
```

Comment:
_I may be missing something. @open-telemetry/go-approvers PTAL_

Is it some hack just to make `internal/internaltest/harness.go` passing?

I think that when using "public" API the `s.stopwatch` will never be `nil`

The same comment applies to `func (s *span) startTime() time.Time`

Reply:
That's true, both checks are for test/benchmark to pass (`harness.go` and `benchmark_test.go`).

Reply:
Can you remove these checks and update the test and benchmark instead?

Reply:
Sure. I will see what I can do.

Reply:
Hi, I traced the relevant logic and discovered that the `stopwatch` actually can be `nil` when the span is not sampled (and in a way is `nil` "by design"). It actually acts as a replacement of `startTime` and would be used to decide whether current span `IsRecording`.

Specifically, in `sdk/trace/span.go` `startSpanInternal` the following check happens before assigning the stopwatch:
```go
	if !isRecording(samplingResult) {
		return span
	}
```

If we try to move the assignment before this check and make stopwatch always non-nil, then in `IsRecording` method:
```go
	return !s.startTime().IsZero() && s.endTime.IsZero()
```
would reports a drop span as isRecording = true.

So my original comment is not correct and should be changed. But maybe we should leave the check as is? What do you say?

Reply:
What do you think of setting some kind of `nullStopwatch` which would return a zero start time when the span is not sampled?

Reply:
Good point. I will make the adjustment and see how it goes.

Reply:
updated